### PR TITLE
feat: add new JournalingBlobWriteSessionConfig usable with gRPC transport

### PR DESCRIPTION
**Pull request by BenWhitehead**
### Overview

New BlobWriteSessionConfig that will use disk to "Journal" data before transmitting to GCS.

By journaling the data to disk we can rewind to an arbitrary offset in the case of failure, while being optimistic and keeping the stream open to gcs.

General flow of data into this new WritableByteChannel:
1. #write(ByteBuffer)
2. Write contents of ByteBuffer to recovery file on disk
3. force a flush to disk
4. transmit contents of ByteBuffer to GCS leaving stream open
5. return

If the stream to gcs is interrupted with a retriable error
1. query the offset that was successfully committed to GCS
2. rewind our transmit context to the offset
3. Open the recovery file at offset from query
4. stream contents from recovery file to GCS leaving stream open
5. Once recovery file contents are transmitted to GCS return

#### Setup
1. c2d-standard-32 in us-east1 
   * with 4 local nvme sdd used for the recovery files 
   * default NIC, premium network tier 
   * debian 11 image
3. regional bucket in us-east1

#### Workload
generate a random file with size `1GiB..2GiB` upload random file to GCS using each of the following configurations:
1. DefaultBlobWriteSessionConfig with 16MiB chunk size
2. DefaultBlobWriteSessionConfig with 64MiB chunk size
4. BufferToDiskThenUpload with the following directories as buffer locations: /mnt/ssd1:/mnt/ssd2:/mnt/ssd3:/mnt/ssd4
5. JournalingBlobWriteSessionConfig with the following directories as journal locations: /mnt/ssd1:/mnt/ssd2:/mnt/ssd3:/mnt/ssd4

Run across `{1,2,4,8,16,32}` concurrent threads to evaluate contention and horizontal scaling.

#### Collected metrics
1. Generate random file
2. record begin instant
3. create new BlobWriteSession
4. BlobWriteSession#open()
6. Copy all bytes from random file to WritableByteChannel from 4
7. close the WritableByteChannel from 4
8. record end instant
9. report `objectSize,UploadStrategy,elapsedTimeUs,Status,ThreadCount`

#### Results summary

Throuhgput in MiB/s, grouped by ThreadCount and UploadStrategy
```
                             count     mean     std     min      50%      75%      90%      99%      max
ThreadCount UploadStrategy
1           PUT 16MiB       4341.0   79.941   8.752  21.599   80.218   85.628   90.710   99.635  106.627
            PUT 64MiB       4341.0  100.410  11.555  20.490  100.022  108.208  115.214  128.251  139.710
            journaling      4341.0  125.820  31.332  45.502  133.590  149.027  161.899  188.716  201.938
2           PUT 16MiB       4237.0   80.000   8.525  15.814   80.693   85.651   90.241   97.958  106.677
            PUT 64MiB       4237.0  101.062  11.030  55.813  101.049  108.007  115.114  127.299  135.149
            journaling      4237.0  125.010  29.761  43.207  131.827  147.362  159.425  182.441  209.000
4           PUT 16MiB       4411.0   79.708   8.357  40.691   80.600   85.567   89.586   95.533  103.506
            PUT 64MiB       4411.0  100.536   9.947  58.084  100.846  107.209  113.144  122.172  131.974
            journaling      4411.0  123.705  30.707  40.082  130.553  147.581  159.995  186.684  222.646
8           PUT 16MiB       4260.0   79.314   8.393   7.148   80.153   85.175   89.319   95.475  100.757
            PUT 64MiB       4260.0   99.913  10.438  60.685  100.450  107.144  112.551  122.409  132.130
            journaling      4260.0  122.931  30.261  42.747  130.306  146.098  158.005  184.798  203.696
16          PUT 16MiB       4473.0   77.735   8.091  24.149   78.483   83.123   87.092   95.740  106.176
            PUT 64MiB       4473.0   97.690   9.987  45.342   97.768  103.996  109.807  122.202  140.906
            journaling      4473.0  118.956  30.486  44.253  122.585  143.344  156.484  182.211  200.777
32          PUT 16MiB       4024.0   72.923   8.045  20.205   73.601   78.575   82.341   88.970  100.665
            PUT 64MiB       4024.0   93.151  10.030  20.913   93.506   99.748  105.297  116.163  128.284
            journaling      4024.0  104.557  30.965  11.785   98.794  129.923  146.747  174.618  200.303
```

### CODE REVIEW

File: google-cloud-storage/src/main/java/com/google/cloud/storage/ChunkSegmenter.java [hunk 1]
```diff
@@ -50,6 +50,10 @@ Hasher getHasher() {
     return hasher;
   }
 
+  ChunkSegment[] segmentBuffer(ByteBuffer bb) {
+    return segmentBuffers(new ByteBuffer[] {bb}, 0, 1);
```
[the remark this answers is not in the document]
Comment (sydney-munro) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/ChunkSegmenter.java:54]: TODO: comment

File: google-cloud-storage/src/main/java/com/google/cloud/storage/JournalingBlobWriteSessionConfig.java [hunk 2]
```diff
@@ -0,0 +1,257 @@
...
+@BetaApi
+@TransportCompatibility(Transport.GRPC)
+public final class JournalingBlobWriteSessionConfig extends BlobWriteSessionConfig {
+  private static final long serialVersionUID = 9059242302276891867L;
+
+  /**
+   * non-final because of {@link java.io.Serializable}, however this field is effectively final as
+   * it is immutable and there is not reference mutator method.
+   */
+  @MonotonicNonNull private transient ImmutableList<Path> paths;
```
[the remark this answers is not in the document]
Comment (sydney-munro) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/JournalingBlobWriteSessionConfig.java:78]: TODO: comment

File: google-cloud-storage/src/main/java/com/google/cloud/storage/ChunkSegmenter.java (hunk 1 above)
[the remark this answers is not in the document]
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/ChunkSegmenter.java:54]: this is an overload of the method `segmentBuffers(ByteBuffer[], offset, length)`

it simplifies the case where only a single ByteBuffer needs to be segmented.

```diff
- chunkSegmenter.segmentBuffer(new ByteBuffer[]{buffer}, 0, 1)
+ chunkSegmenter.segmentBuffer(buffer)
```

File: google-cloud-storage/src/main/java/com/google/cloud/storage/JournalingBlobWriteSessionConfig.java (hunk 2 above)
[the remark this answers is not in the document]
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/JournalingBlobWriteSessionConfig.java:78]: JournalingBlobWriteSessionConfig needs to be java.io.Serializable beacuse it's a field in StorageOptions which is Serializable.

`java.nio.Path` is not Serializable, so we mark it transient and instead serialize the absolute paths as strings for writeObject and readObject.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/RecoveryFile.java [hunk 3]
```diff
@@ -33,25 +35,35 @@
...
   private final ThroughputSink throughputSink;
   private final Runnable onCloseCallback;
 
+  private boolean newFile;
+
   RecoveryFile(Path path, ThroughputSink throughputSink, Runnable onCloseCallback) {
     this.path = path;
     this.throughputSink = throughputSink;
     this.onCloseCallback = onCloseCallback;
+    this.newFile = true;
```
[the remark this answers is not in the document]
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/RecoveryFile.java:58]: An instance of RecoveryFile can be constructed but never used, by not creating the file before it is used we don't leave empty uncleaned files on disk if the constructed instance is abandoned.

A Journaling BlobWriteSession doesn't do anything until `#open()` or `#openAsync()` are called, which allows us to bind the lifecycle of it's RecoveryFile to that of the opened WritableByteChannel.

File: google-cloud-storage/src/main/java/com/google/cloud/storage/SyncAndUploadUnbufferedWritableByteChannel.java [hunk 4]
```diff
@@ -0,0 +1,468 @@
...
+    // resolve current offset and set next
+    long offset = writeCtx.getTotalSentBytes().getAndAdd(contentSize);
+
+    ChecksummedData.Builder checksummedData = ChecksummedData.newBuilder().setContent(b);
+    if (crc32c != null) {
+      checksummedData.setCrc32C(crc32c.getValue());
+    }
+    builder.setWriteOffset(offset).setChecksummedData(checksummedData.build());
+
+    if (!segment.isOnlyFullBlocks()) {
```
[the remark this answers is not in the document]
Comment (BenWhitehead) [reply google-cloud-storage/src/main/java/com/google/cloud/storage/SyncAndUploadUnbufferedWritableByteChannel.java:178]: Correct. If a segment's length does not align with the resumable upload blocksize that must be the final segment.

**Review by sydney-munro — APPROVED** [review]

**Pull request merged**